### PR TITLE
8322980: Debug agent's dumpThread() API should update thread's name before printing it

### PR DESCRIPTION
**Pull request by plummercj**
In threadControl.c, at build time you can decide to keep track of thread names by compiling with "#define DEBUG_THREADNAME". If this is also a debug build, some extra debugging functions are included in the build, including "dumpThread(ThreadNode *node)". These are intended to be called from gdb, or possibly from somewhere in the debug agent implementation, to aid with debugging. When dumpThread() prints the thread's name, it just uses the name that was stored when the thread was created. However, the thread name can change, so dumpThread() should really fetch the current thread name and print it. 

I also added some commented out code to print other useful fields of the ThreadNode, including the ThreadState. These can be enabled by the user as needed.

Tested with all of tier1, and also ran tier2 and tier4 svc tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322980](https://bugs.openjdk.org/browse/JDK-8322980): Debug agent's dumpThread() API should update thread's name before printing it (**Enhancement** - P5)

### Reviewers
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer) ⚠️ Review applies to [eb318d81](https://git.openjdk.org/jdk/pull/17259/files/eb318d81ff9102808651aa12cb3551dd6c98782a)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17259/head:pull/17259` \
`$ git checkout pull/17259`

Update a local copy of the PR: \
`$ git checkout pull/17259` \
`$ git pull https://git.openjdk.org/jdk.git pull/17259/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17259`

View PR using the GUI difftool: \
`$ git pr show -t 17259`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17259.diff">https://git.openjdk.org/jdk/pull/17259.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17259#issuecomment-1876216236)

### CODE REVIEW

**Review by kevinjwalls — APPROVED** [review]
Looks good I think.

A future idea could be to have these ifdefs active in the debug JDK build.  I wouldn't normally run the debugger from the debug build, but if you wanted this extra info it would be more convenient if it was in the debug JDK bundle.

(The changed output might fool some tests so would need checking, but might be worth considering.)

**Comment by plummercj** [comment]
> A future idea could be to have these ifdefs active in the debug JDK build. I wouldn't normally run the debugger from the debug build, but if you wanted this extra info it would be more convenient if it was in the debug JDK bundle.

Possibly DEBUG_THREADNAME should be enabled in debug builds, but I opted not to long ago because of the potential footprint and performance cost. For the `#if 0` code, generally speaking it results in too much output and is really just meant as examples of other things you might want to include in the output, but most cases you would prefer not to have all of it. A possible solution would be to have each line of output controlled by a global variable (either a bitwise flag or a global per line), and then the user could enable additional output at runtime (from within gdb), but I don't feel this is worth the effort.

There aren't any test output concern unless you introduce calls to these API, in which case they are a concern whether or not all the `#if 0` stuff is included.

**Comment by plummercj** [comment]
Could I get one more review please.

File: src/jdk.jdwp.agent/share/native/libjdwp/threadControl.c [hunk 1]
```diff
@@ -2623,15 +2623,58 @@ dumpThreadList(ThreadList *list)
...
+#endif
+
+#if 0
+static jint
+getThreadState(ThreadNode *node)
+{
+  jint state = 0;
+  jvmtiError error = FUNC_PTR(gdata->jvmti,GetThreadState)
+    (gdata->jvmti, node->thread, &state);
+  return state;
```
Comment (sspitsyn) [inline src/jdk.jdwp.agent/share/native/libjdwp/threadControl.c:2654]: Nit: The indent in this file is 4, not 2.

**Review by sspitsyn — APPROVED** [review]
Looks good.
Q: Do you want to keep the debugging code under `#ifdef ... #endif`?

**Comment by plummercj** [comment]
> Looks good. Q: Do you want to keep the debugging code under `#ifdef ... #endif`?

Yes. The "#ifdef DEBUG_THREADNAME" is something that I want the user to have to manually enable at build time to avoid the memory and performance impact. Also, "#if 0" code generates extra output that normally you don't want, but I wanted to leave the code there as examples and so it can be easily enabled when needed."

File: src/jdk.jdwp.agent/share/native/libjdwp/threadControl.c (hunk 1 above)
Comment (plummercj) [reply src/jdk.jdwp.agent/share/native/libjdwp/threadControl.c]: Fixed.

**Review by sspitsyn — APPROVED** [review]

**Comment by plummercj** [comment]
Thank you for the reviews Kevin and Serguei!

/integrate

**Pull request closed without merging**